module/documents/actor/actor.mjs

```diff
@@ -1359,191 +1359,208 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
+    if ( foundry.utils.getType(config) !== "Object" ) {
+      foundry.utils.logCompatibilityWarning(
+        `The \`roll${name}\` method on Actor5e now takes roll, dialog, and message config objects as parameters.`,
+        { since: "DnD5e 4.1", until: "DnD5e 4.5" }
+      );
+      oldFormat = true;
+      const oldConfig = dialog;
+      config = { [type]: config };
+      if ( oldConfig.ability ) config.ability = oldConfig.ability;
+      if ( oldConfig.bonus ) config.bonus = oldConfig.bonus;
+      dialog = {};
+      _applyDeprecatedD20Configs(config, dialog, message, oldConfig);
     }
 
-    // Global ability check bonus.
-    if ( globalBonuses.check ) {
-      parts.push("@checkBonus");
-      data.checkBonus = Roll.replaceFormulaData(globalBonuses.check, data);
+    const skill = this.system.skills?.[config.skill];
+    const skillConfig = CONFIG.DND5E.skills[config.skill];
+    const tool = this.system.tools?.[config.tool];
+    const toolConfig = CONFIG.DND5E.tools[config.tool];
+    if ( ((type === "skill") && !skillConfig) || ((type === "tool") && !toolConfig) ) {
+      return this.rollAbilityTest(config, dialog, message);
     }
 
-    // Ability-specific check bonus.
-    if ( ability?.bonuses.check ) data.abilityCheckBonus = Roll.replaceFormulaData(ability.bonuses.check, data);
-    else data.abilityCheckBonus = 0;
+    const relevant = type === "skill" ? skill : tool;
+    const rollData = this.getRollData();
+    const initialRoll = config.rolls?.pop();
+    const defaultAbility = relevant?.ability ?? (type === "skill" ? skillConfig.ability : toolConfig.ability);
 
-    // Tool-specific check bonus.
-    if ( tool?.bonuses.check || options.bonus ) {
-      parts.push("@toolBonus");
-      const bonus = [];
-      if ( tool?.bonuses.check ) bonus.push(Roll.replaceFormulaData(tool.bonuses.check, data));
-      if ( options.bonus ) bonus.push(Roll.replaceFormulaData(options.bonus, data));
-      data.toolBonus = bonus.join(" + ");
-    }
+    const hasRemarkableAthlete = this.getFlag("dnd5e", "remarkableAthlete")
+      && (game.settings.get("dnd5e", "rulesVersion") === "legacy");
+    const buildConfig = (process, config, formData, index) => {
+      const abilityId = formData?.get("ability") ?? process.ability ?? defaultAbility;
+      const ability = this.system.abilities?.[abilityId];
+
+      let prof = process.prof ?? relevant?.prof;
+      if ( prof && hasRemarkableAthlete ) {
+        const validAbilities = CONFIG.DND5E.characterFlags.remarkableAthlete.abilities;
+        if ( validAbilities.includes(abilityId) && (prof.multiplier <= .5) && (prof.rounding === "down") ) {
+          prof = prof.clone({ multiplier: .5, roundDown: false });
+        } else if ( !validAbilities.includes(abilityId) && (prof.multiplier === .5) && (prof.rounding === "up") ) {
+          if ( this.getFlag("dnd5e", "jackOfAllTrades") ) prof = prof.clone({ roundDown: true });
+          else prof = prof.clone({ multiplier: 0 });
+        }
```

Comment:
I think I understand what we're doing here. Where we try to work out whether remarkable athlete or JoAT has been applied during preparation. But perhaps a bit clearer would be to derive this from the beginning by looking at `_source.skills[config.skill].value` rather than attempting to undo the existing preparation. It could potentially share code with the preparation method.

Reply:
Okay, I was able to go that way but I had to make a tweak to ensure active effects are handled properly. So now during skill/tool preparation it no longer adjusts `skill.value` based on Jack of All Trades & Remarkable Athlete, instead it just takes those flags into account when creating the Proficiency object. This seems fine for our code since we use the proficiency object for determining proficiency, but anyone who was relying on `value` directly might have issues. Not sure that is worth worrying about.

Reply:
Ah right, I hadn't considered Active Effects. It's possible that not updating `value` will cause some downstream disruption. There were a couple of places we were using it in our code judging by this PR, so it's reasonable to assume it's used downstream too.

If we just need a value that is pre-JoAT/RA, but post AE-application, then I would think we can just write it ourselves during preparation, e.g.:

```js
skill.preFeat = skill.value;
skill.value = this.calculateAbilityCheckProficiency(skill.value, skill.ability).multiplier;
```